sdk/eventhub/event-hubs/src/eventHubSender.ts

```diff
@@ -508,24 +500,141 @@ export class EventHubSender extends LinkEntity {
+            );
+          }
+          log.error(err);
+          reject(err);
+        };
+
+        const onModified = (context: EventContext) => {
+          removeListeners();
+          log.error("[%s] Sender '%s', got event modified.", this._context.connectionId, this.name);
+          let err: Error;
+          if (context!.delivery!.remote_state!.error) {
+            err = translate(context!.delivery!.remote_state!.error);
+          } else {
+            err = new Error(
+              `[${this._context.connectionId}] Sender "${this.name}", ` +
+                `received a modified disposition.Hence we are rejecting the promise.`
+            );
+          }
+          log.error(err);
+          reject(err);
+        };
+
+        const removeListeners = (): void => {
+          clearTimeout(waitTimer);
+          // When `removeListeners` is called on timeout, the sender might be closed and cleared
+          // So, check if it exists, before removing listeners from it.
+          if (abortSignal) {
+            abortSignal.removeEventListener("abort", onAborted);
+          }
+          if (this._sender) {
+            this._sender.removeListener(SenderEvents.rejected, onRejected);
+            this._sender.removeListener(SenderEvents.accepted, onAccepted);
+            this._sender.removeListener(SenderEvents.released, onReleased);
+            this._sender.removeListener(SenderEvents.modified, onModified);
+          }
+        };
+
+        const actionAfterTimeout = () => {
+          removeListeners();
+          const desc: string =
+            `[${this._context.connectionId}] Sender "${this.name}" with ` +
+            `address "${this.address}", was not able to send the message right now, due ` +
+            `to operation timeout.`;
+          log.error(desc);
+          const e: Error = {
+            name: "OperationTimeoutError",
+            message: desc
+          };
+          return reject(translate(e));
+        };
+
+        const waitTimer = setTimeout(
+          actionAfterTimeout,
+          getRetryAttemptTimeoutInMs(options.retryOptions)
+        );
+
         if (abortSignal && abortSignal.aborted) {
           // operation has been cancelled, so exit quickly
           return rejectOnAbort();
         }
 
-        let waitTimer: any;
+        if (!this.isOpen()) {
+          log.sender(
+            "Acquiring lock %s for initializing the session, sender and " +
+              "possibly the connection.",
+            this.senderLock
+          );
+
+          try {
+            await defaultLock.acquire(this.senderLock, () => {
+              return this._init();
+            });
+          } catch (err) {
+            err = translate(err);
+            log.error(
+              "[%s] An error occurred while creating the sender %s",
+              this._context.connectionId,
+              this.name,
+              err
+            );
+            reject(err);
```

Comment:
If `init()` fails, then we are calling `reject` here which is good.
But we are not using `return` here.
This would result in the code that follows to be executed which we don't want.

Reply:
Oh, added it .. shouldn't all `reject`s `resolve`s be prepended with return that case ? @chradek

Reply:
reject() and resolve() are callbacks that will ensure that the promise is completed. But they do not stop the flow of control if not used with `return`

Reply:
@ramya0820 
What we talked about before is that a promise won't be fulfilled more than once. So if you call resolve/reject after resolve/reject has already been called once, the value the promise was settled with won't change. However you still need to make sure subsequent code doesn't get ran (which @ramya-rao-a has pointed out)

Reply:
Oh, that makes sense - thanks!